bin/server/cmd/root.go

```diff
@@ -232,162 +213,34 @@ func start() {
 	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
 	go func() {
 		<-c
-		m.l.Info("SIGINT received.")
-		if err := m.Close(); err != nil {
-			fmt.Fprintln(os.Stderr, err)
+		logger.Info("SIGINT received.")
+		if s != nil {
+			_ = s.Shutdown()
 		}
 		cancel()
 	}()
 
-	// Execute program.
-	if err := m.Run(ctx); err != nil {
-		if err != http.ErrServerClosed {
-			m.l.Error(err.Error())
-			m.Close()
-			cancel()
-		}
-	}
-
-	// Wait for CTRL-C.
-	<-ctx.Done()
-}
-
-// NewMain creates a main server based on profile.
-func NewMain(activeProfile server.Profile, logger *zap.Logger) (*Main, error) {
-	fmt.Println("-----Config BEGIN-----")
-	fmt.Printf("mode=%s\n", activeProfile.Mode)
-	fmt.Printf("server=%s:%d\n", host, activeProfile.Port)
-	fmt.Printf("datastore=%s:%d\n", host, activeProfile.DatastorePort)
-	fmt.Printf("frontend=%s:%d\n", frontendHost, frontendPort)
-	fmt.Printf("demoDataDir=%s\n", activeProfile.DemoDataDir)
-	fmt.Printf("readonly=%t\n", readonly)
-	fmt.Printf("demo=%t\n", demo)
-	fmt.Printf("debug=%t\n", debug)
-	fmt.Println("-----Config END-------")
-
-	metadataDB, err := createMetadataDB(&activeProfile, logger)
+	s, err = server.NewServer(ctx, &activeProfile, logger, level)
 	if err != nil {
-		return nil, err
-	}
-
-	return &Main{
-		profile:    &activeProfile,
-		l:          logger,
-		metadataDB: metadataDB,
-	}, nil
-}
-
-func initSetting(ctx context.Context, store *store.Store) (*config, error) {
-	result := &config{}
-	{
-		configCreate := &api.SettingCreate{
-			CreatorID:   api.SystemBotID,
-			Name:        api.SettingAuthSecret,
-			Value:       common.RandomString(secretLength),
-			Description: "Random string used to sign the JWT auth token.",
-		}
-		config, err := store.CreateSettingIfNotExist(ctx, configCreate)
-		if err != nil {
-			return nil, err
+		fmt.Printf("cannot new server, error: %v\n", err)
+		if s != nil {
```

Comment:
Is there any cases we have s != nil and err != nil returned? I think we need better way to handle this situation.

Reply:
SG. I solve this by using defer.